common/lifecyleMgr.go

```diff
@@ -57,12 +64,71 @@ func GetLifecycleMgr() LifecycleMgr {
 
 // single point of control for all outputs
 type lifecycleMgr struct {
-	msgQueue       chan outputMessage
-	progressCache  string // useful for keeping job progress on the last line
-	cancelChannel  chan os.Signal
-	waitEverCalled int32
-	outputFormat   OutputFormat
-	logSanitizer   pipeline.LogSanitizer
+	msgQueue             chan outputMessage
+	progressCache        string // useful for keeping job progress on the last line
+	cancelChannel        chan os.Signal
+	waitEverCalled       int32
+	outputFormat         OutputFormat
+	logSanitizer         pipeline.LogSanitizer
+	inputQueue           chan userInput // msgs from the user
+	allowWatchInput      bool           // accept user inputs and place then in the inputQueue
+	allowCancelFromStdIn bool           // allow user to send in 'cancel' from the stdin to stop the current job
+}
+
+type userInput struct {
+	timeReceived time.Time
+	content      string
+}
+
+// should be started in a single go routine
+func (lcm *lifecycleMgr) watchInputs() {
+	consoleReader := bufio.NewReader(os.Stdin)
+	for {
+		// sleep for a bit, the option might be enabled later
+		if !lcm.allowWatchInput {
+			time.Sleep(time.Microsecond * 500)
+			continue
+		}
+
+		// reads input until the first occurrence of \n in the input,
+		input, err := consoleReader.ReadString('\n')
+		timeReceived := time.Now()
+		if err != nil {
+			continue
+		}
+
+		// remove spaces before/after the content
+		msg := strings.TrimSpace(input)
+
+		if lcm.allowCancelFromStdIn && strings.EqualFold(msg, "cancel") {
+			lcm.cancelChannel <- os.Interrupt
+		} else {
+			lcm.inputQueue <- userInput{timeReceived: timeReceived, content: msg}
+		}
+	}
+}
+
+// get the answer to a question that was asked at a certain time
+// only user input after the specified time is returned to make sure that we are getting the right answer to our question
+func (lcm *lifecycleMgr) getInputAfterTime(time time.Time) string {
```

Comment:
Now here's a thought. What if multiple questions get asked at once? If we're not careful, we could accidentally read another question's answer.

Objects in channels can only be read from a channel once. Consider adding a lock to lcm.Prompt so that only one question ever gets asked at a time.

Reply:
We only ask questions (currently) from with the enumeration code, right?  And that code is single-threaded.

Reply:
We ask questions from `ste` as well, in case of overwrite.

The lifecycleMgr already guarantees that only 1 question get asked at a time because the processors (such `processTextOutput`) outputs the question, and waits until the user answers before doing anything else.